### ISSUE TITLE
Swift5: Binary / File type generated incorrectly as URL in responses

### ISSUE THREAD

**Issue by jenswet-el** [OpenAPITools/openapi-generator#9308]
After some investigation, the test issue when changing the type binary from `URL` to `Data`, is related to the code in ModelUtils line 659.

https://github.com/OpenAPITools/openapi-generator/blob/3894aa47591f71406bed4c8da70bc01b6823023e/modules/openapi-generator/src/main/java/org/openapitools/codegen/DefaultCodegen.java#L2059-L2062

https://github.com/OpenAPITools/openapi-generator/blob/3894aa47591f71406bed4c8da70bc01b6823023e/modules/openapi-generator/src/main/java/org/openapitools/codegen/utils/ModelUtils.java#L643-L660

The binary type will be considered as file type.

To add some interesting bits to the discussion
The Kotlin Client generator, considers binary as BiteArray (or Data in Swift) and file as java.io.File (or URL in Swift)

https://github.com/OpenAPITools/openapi-generator/blob/3894aa47591f71406bed4c8da70bc01b6823023e/modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/AbstractKotlinCodegen.java#L159-L165

The Java Client generator, considers binary as byte[], and file as java.io.File.

Since this is a big change on how files/binary is treated in the Swift Generator that is also a breaking change, and doing things differently from the other generators worries me a bit, so I just would like to be sure that this is the right move, but let's continue this discussion on the new PR..

_Originally posted by @4brunu in https://github.com/OpenAPITools/openapi-generator/pull/9206#discussion_r617754970_

**Comment by jenswet-el** [comment]
@4brunu I've created an issue to discuss the best solution for this. Then we can create a PR based on this.

That is interesting.

I agree this is a breaking change, but nobody will use URL for binary responses in their app. It simply can't work. The URL object is completely useless currently. In our app we "fixed" it by the type-mapping during generation.

Or did I miss something and there is a working use-case for the current implementation?

CC @aymanbagabas

**Comment by 4brunu** [comment]
@jenswet-el thanks for creating this issue.

I have a binary type in swagger file that I consume, that it's used for file upload, and it's working fine with the URL.
I just pass a URL of the file I want to upload, and it works.

```
"File": {
    "type": "string",
    "description": "File to upload",
    "format": "binary"
}
```

**Comment by jenswet-el** [comment]
@4brunu ahhh, you use it as upload. I am talking about download. There it is useless according to our app devs.
```
/image:
    get:
      responses:
        '200':
          description: image in JPEG format
          content:
            image/jpeg:
              schema:
                type: string
                format: binary
```

Generates:

```swift
open class func getImage(apiResponseQueue: DispatchQueue = OpenAPIClientAPI.apiResponseQueue, completion: @escaping ((_ data: Data?, _ error: Error?) -> Void)) {
        getImageWithRequestBuilder().execute(apiResponseQueue) { result -> Void in
            switch result {
            case let .success(response):
                completion(response.body, nil)
            case let .failure(error):
                completion(nil, error)
            }
        }
    }
```

Which maps the binary body into the URL if I am correct?

I don't do Swift (I do the backend), but I got this feedback from our app devs.

**Comment by 4brunu** [comment]
Oh, I see.
Recently the kotlin generator had the same issue that was solved in this PR https://github.com/OpenAPITools/openapi-generator/pull/9284.
Maybe we need to do something similar.

**Comment by aymanbagabas** [comment]
> I have a binary type in swagger file that I consume, that it's used for file upload, and it's working fine with the URL.
> I just pass a URL of the file I want to upload, and it works.

Could you please give us a sample project for this? I never used URLSession nor AlamoFire, but writing back-end Swift in Vapor, a request's body at the end gets converted to `Data` so like for a upload request I would do this using this fork https://github.com/myhealthily/openapi-generator/tree/swift5-vapor

```swift
app.post() { request -> Response in
    let file = Data(base64Encoded: base64)!
    PetAPI.uploadFile(petId: 99, file: file).whenComplete { result in
        switch result {
        case .success(let uploadFile):
            switch uploadFile {
            case .http0(let value, _), .http200(let value, _):
                print(value.code, value.type, value.message)
            }
        case .failure(let error):
            print(error)
        }
    }
    return Response(status: .ok)
}
```

[petstore.zip](https://github.com/OpenAPITools/openapi-generator/files/6393235/petstore.zip)

**Comment by 4brunu** [comment]
Sure, here you have a sample project that show's the API usage.
https://github.com/OpenAPITools/openapi-generator/blob/5d946289efe1cd69fc680ab5efc16ac37a02a7db/samples/client/petstore/swift5/default/SwaggerClientTests/SwaggerClientTests/PetAPITests.swift#L65-L89

And here you have the URLSession implementation for file upload
https://github.com/OpenAPITools/openapi-generator/blob/5d946289efe1cd69fc680ab5efc16ac37a02a7db/samples/client/petstore/swift5/default/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift#L424-L566

**Comment by aymanbagabas** [comment]
@4brunu URLSessionImplementations.swift translates the URLs into `Data` when constructing the Request. This doesn't mean that `file` and `binary` are equivalent to `URL` in Swift`, instead, here `URLSessionImplementations.swift` and `FileUtils.swift` are handling the translation of URLs to Data. At the end, the data is grabbed via `Data(contentsOf:)` in `let fileData = try Data(contentsOf: fileURL)` and then used in `body.append(fileData)` the type here is `Data` *not* `URL`.

IMO, this should be the job of the user to grab the raw data from URLs and not the client library. We could have helpers to assist with that but *not* to make that the default behavior.

CC @jenswet-el @wing328 

https://github.com/OpenAPITools/openapi-generator/blob/5d946289efe1cd69fc680ab5efc16ac37a02a7db/samples/client/petstore/swift5/default/PetstoreClient/Classes/OpenAPIs/URLSessionImplementations.swift#L491-L524

**Comment by 4brunu** [comment]
`URL` is only kind of a "pointer" to the `Data`, so to send it to the server, it's necessary to convert the `URL` to `Data`.
Maybe this transition can be done with a flag to ensure backward compatible and in a future version, we can remove the old behaviour.
The only thing that is holding this back to me, is that the other generators, also use `File` instead of `ByteArray`, just like the Swift Generator does, so I want to have 100% sure that this is the right path.

**Comment by 4brunu** [comment]
I have looked into multiple generators, but the only conclusion that I got is that there is no uniformity between them.

So maybe we can go forward with this change, and create a feature flag to avoid the breaking change.

what do you think of this?

c#
```
typeMapping.put("binary", "byte[]");
typeMapping.put("ByteArray", "byte[]");
typeMapping.put("file", "System.IO.Stream");
```

java
```
typeMapping.put("file", "File");
```

java play framework
```
typeMapping.put("file", "InputStream");
```

Ruby
```
typeMapping.put("file", "File");
typeMapping.put("binary", "String");
typeMapping.put("ByteArray", "String");
```

Kotlin
```
typeMapping.put("ByteArray", "kotlin.ByteArray");
typeMapping.put("file", "java.io.File");
typeMapping.put("binary", "kotlin.ByteArray");
```

Dart
```
typeMapping.put("file", "MultipartFile");
typeMapping.put("binary", "MultipartFile");
typeMapping.put("ByteArray", "String");
```

TypeScript
```
typeMapping.put("binary", "any");
typeMapping.put("File", "any");
typeMapping.put("file", "any");
typeMapping.put("ByteArray", "string");
```

**Comment by 4brunu** [comment]
@aymanbagabas do you want to create a PR with this change?

**Comment by aymanbagabas** [comment]
@4brunu I've created https://github.com/OpenAPITools/openapi-generator/pull/9419 and rebased https://github.com/OpenAPITools/openapi-generator/pull/8515 on top of #9419

**Comment by 4brunu** [comment]
After looking at this issue once more, I found the following code.

https://github.com/OpenAPITools/openapi-generator/blob/163d13e4159e695a4bea8d428e860e121c4c0234/samples/client/petstore/swift5/urlsessionLibrary/Sources/URLSessionImplementations.swift#L196-L229

Looking at this, the binary download should have always worked.
The url that is returned as a response, is a url to a local file with the binary content return by the server.
I don't have any endpoint to test this.
@aymanbagabas @jenswet-el Do you have any endpoint where you could test this?
Thanks

**Comment by aymanbagabas** [comment]
> @aymanbagabas @jenswet-el Do you have any endpoint where you could test this?

You can use [HTTP Cats](https://http.cat/) API, https://http.cat/500.jpg for instance

**Comment by 4brunu** [comment]
Do they have a swagger file?

**Comment by aymanbagabas** [comment]
> Do they have a swagger/openapi file?

No, but it's a simple API. Here's one:

```
openapi: 3.0.0
info:
  title: HTTP Cats
  description: Cats
  version: 1.0.0
servers:
  - url: https://http.cat/
    description: Cats
paths:
  /{code}:
    get:
      summary: Get Internal Server Error
      parameters:
        - in: path
          name: code
          schema:
            type: integer
          required: true
          description: HTTP Code
      responses:
        '200':
          description: Cat image
          content:
            image/jpeg:
              schema:
                type: string
                format: binary

```

**Comment by 4brunu** [comment]
Thanks.
After some tests, I confirm that the download of binary response was not working.
I opened the following PR to fix it https://github.com/OpenAPITools/openapi-generator/pull/9595.

**Comment by 4brunu** [comment]
Now that https://github.com/OpenAPITools/openapi-generator/pull/9595 got merged, the type `binary` mapped to `URL` work with both upload and download.
When type `binary` is mapped to `Data` only the download works, the file upload doesn't work (I think looking at the code, but I haven't tested it.)
Yesterday I tried to fix the file upload when mapped to `Data`, but I faced some challenges, because there are some info that is harder to get from `Data` than from a `URL`.
For example:
- What should be the file name and file extension?
- It's harder to extract the mimeType from `Data` than from a `URL`. The only way that I found to do this was something like https://stackoverflow.com/a/32765708/976628 or https://github.com/sendyhalim/Swime but that way we would have a limited set of binary type support. 

That being said, now that the `binary` mapped to `URL` works in all the cases, is the switching to `Data` mapping worth pursuing?
